File: 2018/src/main/scala/advent/solutions/Day11.scala
package advent.solutions

import advent.util.Util

object Day11 {

  def part1(lines: Array[String]): (Int, Int) = {
    val serialNumber = lines(0).toInt
    val grid = Range(0, 301).foldLeft(Map[(Int, Int), Int]())((grid, i) =>
      Range(0, 301).foldLeft(grid)((grid, j) => grid.updated((i, j), calculatePower(i, j, serialNumber))))
    Range(0, 298).foldLeft(((0, 0), Int.MinValue))({ case ((position, max), i) =>
      Range(0, 298).foldLeft((position, max))((bestPosition, j) => {
        val positionValue = grid(i, j) + grid(i + 1, j) + grid(i + 2, j) +
          grid(i, j + 1) + grid(i + 1, j + 1) + grid(i + 2, j + 1) +
          grid(i, j + 2) + grid(i + 1, j + 2) + grid(i + 2, j + 2)
        if (positionValue > bestPosition._2) ((i, j), positionValue)
        else bestPosition
      })
    })._1
  }

  def part2(lines: Array[String]): (Int, Int, Int) = {
    val serialNumber = lines(0).toInt
    val grid = Range(0, 301).foldLeft(Map[(Int, Int), Int]())((grid, i) => {
      Range(0, 301).foldLeft(grid)((grid, j) => grid.updated((i, j), calculatePower(i, j, serialNumber)))
    })
    Range(0, 301).foldRight(Map[(Int, Int, Int), Int]())((i, results) =>
      Range(0, 301).foldRight(results)((j, results) =>
        Range(1, Math.min(300 - i, 300 - j) + 2).foldLeft(results)((results, size) => {
          var positionValue = grid(i, j)
          if (size != 1)
            positionValue += results(i + 1, j + 1, size - 1) +
              Range(1, size).foldLeft(0)((value, position) => value + grid(i + position, j) + grid(i, j + position))
          results.updated((i, j, size), positionValue)
        })
      )).maxBy(_._2)._1
  }

  def calculatePower(i: Int, j: Int, serialNumber: Int): Int = {
    val rackId = i + 10
    ((rackId * j + serialNumber) * rackId % 1000) / 100 - 5
  }

  def main(args: Array[String]): Unit = {
    println("Part 1 result: " + part1(Util.readFileLines(args(0))))
    println("Part 2 result: " + part2(Util.readFileLines(args(1))))
  }

}

File: 2019/src/main/kotlin/hr/dodomix/advent/solutions/Day11.kt
package hr.dodomix.advent.solutions

import hr.dodomix.advent.util.ProcessorUtil.constructMemory
import hr.dodomix.advent.util.ProcessorUtil.hasProgramFinished
import hr.dodomix.advent.util.ProcessorUtil.runProgram
import java.math.BigInteger

class Day11 {

    fun dayDirectory() = "day11"

    fun part1(input: List<String>): Int {
        val memory = constructMemory(input)
        val grid = mutableMapOf<Pair<Int, Int>, Int>().withDefault { 0 }
        var position = Pair(0, 0)
        val coloredPositions = mutableSetOf<Pair<Int, Int>>()
        var direction = 0
        while (!hasProgramFinished(memory)) {
            val output = mutableListOf<BigInteger>()
            runProgram(
                memory,
                input = listOf(BigInteger.valueOf(grid.getValue(position).toLong())),
                output = output,
                waitForInput = true
            )
            val newColor = output[0].toInt()
            if (newColor != grid.getValue(position)) {
                grid[position] = newColor
                coloredPositions.add(position)
            }
            val directionChange = output.get(1).toInt()
            if (directionChange == 0) {
                direction--
                direction += 4
                direction %= 4
            } else {
                direction++
                direction += 4
                direction %= 4
            }
            position = when (direction) {
                0 -> Pair(position.first - 1, position.second)
                1 -> Pair(position.first, position.second + 1)
                2 -> Pair(position.first + 1, position.second)
                3 -> Pair(position.first, position.second - 1)
                else -> throw RuntimeException("Invalid direction")
            }
        }
        return coloredPositions.size
    }

    fun part2(input: List<String>): Int {
        val memory = constructMemory(input)
        val grid = mutableMapOf(Pair(Pair(0, 0), 1)).withDefault { 0 }
        var position = Pair(0, 0)
        var direction = 0
        while (!hasProgramFinished(memory)) {
            val output = mutableListOf<BigInteger>()
            runProgram(
                memory,
                input = listOf(BigInteger.valueOf(grid.getValue(position).toLong())),
                output = output,
                waitForInput = true
            )
            grid[position] = output[0].toInt()
            val directionChange = output.get(1).toInt()
            if (directionChange == 0) {
                direction--
                direction += 4
                direction %= 4
            } else {
                direction++
                direction += 4
                direction %= 4
            }
            position = when (direction) {
                0 -> Pair(position.first - 1, position.second)
                1 -> Pair(position.first, position.second + 1)
                2 -> Pair(position.first + 1, position.second)
                3 -> Pair(position.first, position.second - 1)
                else -> throw RuntimeException("Invalid direction")
            }
        }

        for (row in grid.keys.map { it.first }.min()!!..grid.keys.map { it.first }.max()!!) {
            for (column in grid.keys.map { it.second }.min()!!..grid.keys.map { it.second }.max()!!) {
                print(if (grid.getValue(Pair(row, column)) == 1) '#' else ' ')
            }
            println()
        }

        return 0
    }
}
